tests/integration/acp.go

```diff
@@ -69,12 +113,301 @@ func addPolicyACP(
+	cfg.Set("consensus.timeout_propose", "0.5s")
+	cfg.Set("consensus.timeout_commit", "1s")
+
+	_, err = cfg.WriteTo(fo)
+	if err != nil {
+		return nil, err
+	}
+	err = fo.Close()
+	if err != nil {
+		return nil, err
+	}
+
+	out, err = exec.Command(
+		"sourcehubd", "keys", "import-hex", validatorName, acpKeyHex,
+		"--keyring-backend", keyringBackend,
+		"--home", directory,
+	).CombinedOutput()
+	s.t.Log(string(out))
+	if err != nil {
+		return nil, err
+	}
+
+	out, err = exec.Command(
+		"sourcehubd", "keys", "show", validatorName,
+		"--address",
+		"--keyring-backend", keyringBackend,
+		"--home", directory,
+	).CombinedOutput()
+	s.t.Log(string(out))
+	if err != nil {
+		return nil, err
+	}
+
+	// The result is suffexed with a newline char so we must trim the whitespace
+	validatorAddress := strings.TrimSpace(string(out))
+	s.sourcehubAddress = validatorAddress
+
+	out, err = exec.Command(
+		"sourcehubd", "genesis", "add-genesis-account", validatorAddress, "900000000stake",
+		"--keyring-backend", keyringBackend,
+		"--home", directory,
+	).CombinedOutput()
+	s.t.Log(string(out))
+	if err != nil {
+		return nil, err
+	}
+
+	out, err = exec.Command(
+		"sourcehubd", "genesis", "gentx", validatorName, "10000000stake",
+		"--chain-id", chainID,
+		"--keyring-backend", keyringBackend,
+		"--home", directory,
+	).CombinedOutput()
+	s.t.Log(string(out))
+	if err != nil {
+		return nil, err
+	}
+
+	out, err = exec.Command("sourcehubd", "genesis", "collect-gentxs", "--home", directory).CombinedOutput()
+	s.t.Log(string(out))
+	if err != nil {
+		return nil, err
+	}
+
+	// We need to lock across all the test processes as we assign ports to the source hub instance as this
+	// process involves finding free ports, dropping them, and then assigning them to the source hub node.
+	//
+	// We have to do this because source hub (cosmos) annoyingly does not support automatic port assignment
+	// (appart from the p2p port which we just managage here for consistency).
+	//
+	// We need to lock before getting the ports, otherwise they may try and use the port we use for locking.
+	// We can only unlock after the source hub node has started and begun listening on the assigned ports.
+	unlock, err := crossLock(55555)
+	if err != nil {
+		return nil, err
+	}
+	defer unlock()
+
+	gRpcPort, releaseGrpcPort, err := getFreePort()
+	if err != nil {
+		return nil, err
+	}
+
+	rpcPort, releaseRpcPort, err := getFreePort()
+	if err != nil {
+		return nil, err
+	}
+
+	p2pPort, releaseP2pPort, err := getFreePort()
+	if err != nil {
+		return nil, err
+	}
+
+	pprofPort, releasePprofPort, err := getFreePort()
+	if err != nil {
+		return nil, err
+	}
+
+	gRpcAddress := fmt.Sprintf("127.0.0.1:%v", gRpcPort)
+	rpcAddress := fmt.Sprintf("tcp://127.0.0.1:%v", rpcPort)
+	p2pAddress := fmt.Sprintf("tcp://127.0.0.1:%v", p2pPort)
+	pprofAddress := fmt.Sprintf("127.0.0.1:%v", pprofPort)
+
+	releaseGrpcPort()
+	releaseRpcPort()
+	releaseP2pPort()
+	releasePprofPort()
```

Comment:
question: shouldn't this deferred?

Reply:
No, as the port would still be in use when `sourcehubd start` tries to use it :)